image_proc/src/crop_decimate.cpp

```diff
@@ -0,0 +1,315 @@
+    decimation_x /= 2;
+    decimation_y /= 2;
+  }
+
+  // Apply further downsampling, if necessary
+  if (decimation_x > 1 || decimation_y > 1)
+  {
+    cv::Mat decimated;
+
+    if (interpolation_ == image_proc::CropDecimate_NN)
+    {
+      // Use optimized method instead of OpenCV's more general NN resize
+      int pixel_size = output.image.elemSize();
+      switch (pixel_size)
+      {
+        // Currently support up through 4-channel float
+        case 1:
+          decimate<1>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 2:
+          decimate<2>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 3:
+          decimate<3>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 4:
+          decimate<4>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 6:
+          decimate<6>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 8:
+          decimate<8>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 12:
+          decimate<12>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        case 16:
+          decimate<16>(output.image, decimated, decimation_x, decimation_y);
+          break;
+        default:
+          RCLCPP_ERROR(get_logger(),
+            "Unsupported pixel size, %d bytes", pixel_size);
+          return;
+      }
+    }
+    else
+    {
+      // Linear, cubic, area, ...
+      cv::Size size(output.image.cols / decimation_x, output.image.rows / decimation_y);
+      cv::resize(output.image, decimated, size, 0.0, 0.0, interpolation_);
+    }
+
+    output.image = decimated;
+  }
+
+  // Create output Image message
+  /// @todo Could save copies by allocating this above and having output.image alias it
+  sensor_msgs::msg::Image::SharedPtr out_image = output.toImageMsg();
+
+  // Create updated CameraInfo message
+  sensor_msgs::msg::CameraInfo::SharedPtr out_info =
+    std::make_shared<sensor_msgs::msg::CameraInfo>(*info_msg);
+  int binning_x = std::max((int)info_msg->binning_x, 1);
+  int binning_y = std::max((int)info_msg->binning_y, 1);
+  out_info->binning_x = binning_x * decimation_x_;
+  out_info->binning_y = binning_y * decimation_y_;
+  out_info->roi.x_offset += offset_x_ * binning_x;
+  out_info->roi.y_offset += offset_y_ * binning_y;
+  out_info->roi.height = height * binning_y;
+  out_info->roi.width = width * binning_x;
+  // If no ROI specified, leave do_rectify as-is. If ROI specified, set do_rectify = true.
+  if (width != (int)image_msg->width || height != (int)image_msg->height)
+    out_info->roi.do_rectify = true;
+
+  if (!target_frame_id_.empty()) {
+    out_image->header.frame_id = target_frame_id_;
+    out_info->header.frame_id = target_frame_id_;
+  }
+
+  pub_.publish(out_image, out_info);
```

Comment:
For Dashing, it is recommended to publish the pointers to these if we aren't going to use `unique_ptr`s.

Reply:
Those are both pointers

Reply:
See deprecation of `shared_ptr` publish signatures here: https://github.com/ros2/rclcpp/pull/709 Discussion about preferring `unique_ptr`: https://github.com/ros2/demos/pull/327 (since it works whether you're using intra-process or not and is advantageous when you are).

Reply:
Sure, and that was part of my hot patch, but this isn't a rclcpp publisher, its a image transport publisher that wants a shared pointer specifically 

https://github.com/ros-perception/image_common/blob/dashing/image_transport/include/image_transport/camera_publisher.h#L114

Reply:
You can use refs, but, IMO, `unique_ptr` + `std::move` is the most efficient and flexible.

Reply:
HUH. OK, nevermind then. Seems an odd choice given the way that the core is going, but whatevs.